README.md

[![Arduino CI](https://github.com/RobTillaart/DS18B20_INT/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/DS18B20_INT/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/DS18B20_INT/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/DS18B20_INT/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/DS18B20_INT/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/DS18B20_INT/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/DS18B20_INT.svg?maxAge=3600)](https://github.com/RobTillaart/DS18B20_INT/releases)


# DS18B20_INT

Minimalistic library for the DS18B20 temperature sensor - restricted to one sensor per pin.


## Arduino Temperature Control Library (ATCL)

This DS18B20 library is a minimalistic library for a DS18B20 sensor.
It will give only temperatures in whole degrees C.
Goal is to minimize footprint.

If you need more functions or control over the DS18B20 family I refer to the library
of Miles Burton - https://github.com/milesburton/Arduino-Temperature-Control-Library

I'm a great fan of the above library however some time ago I needed to strip it down 
to save a few dozen bytes. I reworked that minimalistic version into a library and I 
added a number of Arduino examples to help you get started.


#### Related

This library is related to 
- https://github.com/RobTillaart/DS18B20_INT
- https://github.com/RobTillaart/DS18B20_RT
- https://github.com/milesburton/Arduino-Temperature-Control-Library


## Interface

```cpp
#include "DS18B20_INT.h"
```


#### Core

This DS18B20_INT library supports only the DS18B20, only one sensor per pin, no parasite 
mode, no Fahrenheit and no alarm functions. The only feature the class supports is 
the asynchronous reading of the temperature by means of three core functions:

- **DS18B20_INT(OneWire \*)** constructor needs a reference to OneWire object.
- **bool begin(uint8_t retries = 3)** resets oneWire and set resolution default to 9 bit.  
Returns true if address / device is found and all is OK. 
There will be a number of retries to connect, default 3.
- **bool isConnected()** Returns true if address / device is found.
- **void requestTemperatures()** trigger temperature conversion.
- **bool isConversionComplete()** check if conversion is complete.
- **int16_t getTempC()** returns temperature in whole degrees only. -55..125  
or  -127 = DEVICE_DISCONNECTED
- **bool getAddress()** returns true if the sensor is configured (available).


#### CentiC part

The following functions are experimental since 0.2.0 and not tested a lot by me.
They allow to use a higher resolution while not using floats. 
Goal is to keep the footprint of the library small.

- **bool setResolution(uint8_t bits = 9)** sets the internal resolution to 9, 10, 11 or 12 bits. 
Other numbers will be mapped on 9. 
This will affect the conversion time for a measurement.
Internally it will call **begin()** to set the new resolution.
Returns false if no address / device can be found.
- **void getResolution()** returns the bits set, default 9.
Convenience function.
- **getTempCentiC(void)** returns the measured temperature times 100. -5500..12500
So 10.62°C will be returned as 1062.
Note one might need to set the resolution.


## Operation

This library supports only one DS18B20 per Arduino/ MCU pin.

```
    //  BOTTOM VIEW
    //
    //          PIN  MEANING
    //   /---+
    //  /  o |  1    GND
    //  |  o |  2    DATA
    //  \  o |  3    VCC
    //   \---+
```
(always check datasheet)

Connect a pull-up resistor 4.7 KOhm between pin3 and pin2. 
When the wires are longer this resistor needs to be smaller.


#### -127 and 85

Two specific return values from reading the sensor:

- minus 127 == DEVICE_DISCONNECTED
- plus 85 is the power on default. 
If you get this unexpected it may indicate a power problem


#### Pull up resistor

An **indicative** table for pull up resistors, (E12 series), to get started.

Note: thicker wires require smaller resistors (typically 1 step in E12 series) 


|  Length         |   5.0 Volt  |  3.3 Volt  |  Notes  |
|----------------:|------------:|-----------:|:--------|
|  10cm (4")      |    10K0     |    6K8     |  might work without  |
|  20cm (8")      |     8K2     |    4K7     |
|  50cm (20")     |     4K7     |    3K3     | 
|  100cm (3'4")   |     3K3     |    2K2     | 
|  200cm (6'8")   |     2K2     |    1K0     | 
|  500cm (16'8")  |     1K0     |    \*      |  
|  longer         |     \*      |    \*      |

\* = no info, smaller?


## Performance

To elaborate connected state.

Tested on UNO, 16 MHz, time in microseconds.
Measured with DS18B20_performance.ino

|  function              |  not connected  |  connected  |
|:-----------------------|:---------------:|:-----------:|
|  begin                 |            920  |      -      |
|  getAddress            |              4  |      -      |
|  requestTemperatures   |           1276  |      -      |
|  isConversionComplete  |             72  |      -      |
|  getTempC              |            912  |      -      |


## Credits

Miles Burton who originally developed the Arduino Temperature Control Library.
and all people who contributed to that lib.


## Future

#### Must

- elaborate performance connected state.
- document 85 and -127 better
  - #define DEVICE_POWER_ON       85
  - example reread?

#### Should

- better error handling
  - DS18B20_ERR_DISCONNECT
  - DS18B20_OK
  - DS18B20_ERR_POWER_ON.
  - DS18B20_ERR_TIMEOUT ?

#### Could

- add rounding for **getTempC()**.
  - now it truncates, so it can be 0.5°C off.
  - add "0.5" to raw and truncate improves only for 10 bits and higher.
  - ==> in conflict with minimalistic goal?

#### Wont

- unit tests
  - get it working is too time consuming.





CHANGELOG.md
# Change Log DS18B20_INT

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.0] - 2023-03-05
- fix #15 infinite loop
- move sources to src folder to comply with PlatformIO dependency system. 
- make oneWire dependency explicit in library.properties
- update readme.md
- fix changelog.md


----

## [0.2.2] - 2023-02-03
- improve disconnect detection
- add **isConnected(retries)**
- change signature **bool setResolution(bits)** to see if device is still available.
- update GitHub actions
- update license 2023
- add example DS18B20_performance.ino (start not complete)
- add example DS18B20_two_sensor.ino
- add example oneWireSearch.ino
- update readme.md

## [0.2.1] - 2022-11-02
- add changelog.md
- add rp2040 to build-CI

## [0.2.0] - 2022-06-23
- fix #10 getTempCentiC() (thanks negroKiordi)
- fix reading sensor

----

## [0.1.7] - 2021-12-17
- update library.json
- update license
- minor edits

## [0.1.6] - 2021-10-03
- add dependency
- fix build-CI

## [0.1.5] - 2021-06-16
- add retries parameter to begin()

## [0.1.4] - 2021-05-26
- add OneWire.reset() to begin()


## [0.1.3] - 2020-12-20
- add Arduino-CI + unit test

## [0.1.2] - 2020-08-05
- refactor
- sync with DS18B20

## [0.1.1] - 2019-??-??

## [0.1.0] - 2017-07-25
- initial version





test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-03
// PURPOSE: unit tests for the DS18B20_INT library
//          https://github.com/RobTillaart/DS18B20_INT
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)

#include <ArduinoUnitTests.h>

#define assertEqualFloat(arg1, arg2, arg3)  assertOp("assertEqualFloat", "expected", fabs(arg1 - arg2), compareLessOrEqual, "<=", "actual", arg3)
#define assertEqualINF(arg)  assertOp("assertEqualINF", "expected", INFINITY, compareEqual, "==", "actual", arg)
#define assertEqualNAN(arg)  assertOp("assertEqualNAN", "expected", true, compareEqual, "==", "actual", isnan(arg))


#include "Arduino.h"
#include "DS18B20_INT.h"


unittest_setup()
{
}

unittest_teardown()
{
}

unittest(test_constructor)
{
  OneWire oneWire(ONE_WIRE_BUS);
  DS18B20_INT sensor(&oneWire);
  
  sensor.requestTemperatures();
  delay(1000);
  assertFalse(sensor.isConversionComplete());
  
  assertEqual(0, sensor.getTempC());
}

unittest_main()

// --------


src/DS18B20_INT.h
#pragma once
//
//    FILE: DS18B20_INT.h
//  AUTHOR: Rob.Tillaart@gmail.com
// VERSION: 0.3.0
//    DATE: 2017-07-25
// PUPROSE: Minimalistic library for DS18B20 temperature sensor
//          uses only integer math (no float to minimize footprint)
//     URL: https://github.com/RobTillaart/DS18B20_INT
//          https://github.com/RobTillaart/DS18B20_RT

//
//          BOTTOM VIEW
//
//                  PIN     MEANING
//           /---+
//          /  o |  1       GND
//          |  o |  2       DATA
//          \  o |  3       VCC
//           \---+
//


#define DS18B20_INT_LIB_VERSION       (F("0.3.0"))

#include "Arduino.h"
#include "OneWire.h"

//  Error Code
#define DEVICE_DISCONNECTED           -127


typedef uint8_t DeviceAddress[8];


class DS18B20_INT
{
public:
  explicit  DS18B20_INT(OneWire * ow);
  bool      begin(uint8_t retries = 3);
  bool      isConnected(uint8_t retries = 3);

  void      requestTemperatures(void);
  int16_t   getTempC(void);
  bool      isConversionComplete(void);
  bool      getAddress(uint8_t* buf);

  bool      setResolution(uint8_t bits = 9);
  uint8_t   getResolution();  //  returns cached value

  int16_t   getTempCentiC(void);


private:
  DeviceAddress _deviceAddress;
  OneWire*      _oneWire;
  bool          _addressFound;

  uint8_t       _resolution;
  int16_t       _readRaw();
  bool          _setResolution();
};


//  -- END OF FILE --



src/DS18B20_INT.cpp
//
//    FILE: DS18B20_INT.cpp
//  AUTHOR: Rob.Tillaart@gmail.com
// VERSION: 0.3.0
//    DATE: 2017-07-25
// PUPROSE: library for DS18B20 temperature sensor - integer only.
//     URL: https://github.com/RobTillaart/DS18B20_INT
//          https://github.com/RobTillaart/DS18B20_RT


#include "DS18B20_INT.h"


//  OneWire commands
#define STARTCONVO              0x44
#define READSCRATCH             0xBE
#define WRITESCRATCH            0x4E


//  Device resolution
#define TEMP_9_BIT              0x1F    //   9 bit
#define TEMP_10_BIT             0x3F    //  10 bit
#define TEMP_11_BIT             0x5F    //  11 bit
#define TEMP_12_BIT             0x7F    //  12 bit


DS18B20_INT::DS18B20_INT(OneWire* ow)
{
  _oneWire      = ow;
  _addressFound = false;
  _resolution   = TEMP_9_BIT;
}


bool DS18B20_INT::begin(uint8_t retries)
{
  if (isConnected(retries))
  {
    _setResolution();
  }
  return _addressFound;
}


bool DS18B20_INT::isConnected(uint8_t retries)
{
  _addressFound = false;
  for (uint8_t rtr = retries; (rtr > 0) && (_addressFound == false); rtr--)
  {
    _oneWire->reset();
    _oneWire->reset_search();
    _deviceAddress[0] = 0x00;
    _oneWire->search(_deviceAddress);
    _addressFound = _deviceAddress[0] != 0x00 &&
                _oneWire->crc8(_deviceAddress, 7) == _deviceAddress[7];
  }
  return _addressFound;
}


void DS18B20_INT::requestTemperatures(void)
{
  _oneWire->reset();
  _oneWire->skip();
  _oneWire->write(STARTCONVO, 0);
}


bool DS18B20_INT::isConversionComplete(void)
{
  return (_oneWire->read_bit() == 1);
}


int16_t DS18B20_INT::getTempC(void)
{
  if (isConnected(3) == false)
  {
    return DEVICE_DISCONNECTED;
  }
  int16_t rawTemperature = _readRaw();
  rawTemperature >>= 4;
  if (rawTemperature < -55)
  {
    return DEVICE_DISCONNECTED;
  }
  return rawTemperature;
}


bool DS18B20_INT::getAddress(uint8_t* buf)
{
  if (_addressFound)
  {
    for (uint8_t i = 0; i < 8; i++)
    {
      buf[i] = _deviceAddress[i];
    }
  }
  return _addressFound;
}


bool DS18B20_INT::setResolution(uint8_t bits)
{
  if (isConnected())
  {
    switch (bits)
    {
      case 12: _resolution = TEMP_12_BIT;  break;
      case 11: _resolution = TEMP_11_BIT;  break;
      case 10: _resolution = TEMP_10_BIT;  break;
      default: _resolution = TEMP_9_BIT;   break;
    }
    _setResolution();
  }
  return _addressFound;
}


uint8_t DS18B20_INT::getResolution()
{
  switch (_resolution)
  {
    case TEMP_12_BIT: return 12;
    case TEMP_11_BIT: return 11;
    case TEMP_10_BIT: return 10;
    case TEMP_9_BIT:  return 9;
  }
  return 0;
}


int16_t DS18B20_INT::getTempCentiC(void)
{
  if (isConnected(3) == false)
  {
    return DEVICE_DISCONNECTED;
  }
  int16_t rawTemperature = _readRaw();
  //  rawTemperature = rawTemperature * 100 / 16;
  rawTemperature *= 25;
  rawTemperature >>= 2;
  //  use at own risk. (not tested)
  if (rawTemperature < -5500)
  {
    return DEVICE_DISCONNECTED * 100;
  }
  return rawTemperature;
}


//////////////////////////////////////////////////
//
//  PRIVATE
//
int16_t DS18B20_INT::_readRaw(void)
{
  _oneWire->reset();
  _oneWire->select(_deviceAddress);
  _oneWire->write(READSCRATCH);
  int16_t rawTemperature = ((int16_t)_oneWire->read());
  rawTemperature |= _oneWire->read() << 8;
  _oneWire->reset();
  return rawTemperature;
}


bool DS18B20_INT::_setResolution()
{
  if (_addressFound)
  {
    _oneWire->reset();
    _oneWire->select(_deviceAddress);
    _oneWire->write(WRITESCRATCH);
    //  two dummy values for LOW & HIGH ALARM
    _oneWire->write(0);
    _oneWire->write(100);
    //  lowest as default as we do only integer math.
    _oneWire->write(_resolution);
    _oneWire->reset();
  }
  return _addressFound;
}


//  -- END OF FILE --

